src/tensors/levels/atomiclevel.jl

```diff
@@ -0,0 +1,250 @@
+@inline level_axes(lvl::AtomicLevel{AVal, Lvl}) where {AVal, Lvl} = level_axes(lvl.lvl)
+@inline level_eltype(::Type{AtomicLevel{AVal, Lvl}}) where {AVal, Lvl} = level_eltype(Lvl)
+@inline level_default(::Type{<:AtomicLevel{AVal, Lvl}}) where {AVal, Lvl} = level_default(Lvl)
+
+# FIXME: These.
+(fbr::Tensor{<:AtomicLevel})() = SubFiber(fbr.lvl, 1)()
+(fbr::SubFiber{<:AtomicLevel})() = fbr #TODO this is not consistent somehow
+function (fbr::SubFiber{<:AtomicLevel})(idxs...)
+    return Tensor(fbr.lvl)(idxs...)
+end
+
+countstored_level(lvl::AtomicLevel, pos) = countstored_level(lvl.lvl, pos)
+
+mutable struct VirtualAtomicLevel <: AbstractVirtualLevel
+    lvl # the level below us.
+    ex
+    locks
+    Tv
+    Val
+    AVal
+    Lvl
+end
+postype(lvl:: AtomicLevel) = postype(lvl.lvl)
+
+postype(lvl:: VirtualAtomicLevel) = postype(lvl.lvl)
+
+is_level_injective(lvl::VirtualAtomicLevel, ctx) = [is_level_injective(lvl.lvl, ctx)..., true]
+is_level_concurrent(lvl::VirtualAtomicLevel, ctx) = [is_level_concurrent(lvl.lvl, ctx)..., true]
+is_level_atomic(lvl::VirtualAtomicLevel, ctx) = true
+
+function lower(lvl::VirtualAtomicLevel, ctx::AbstractCompiler, ::DefaultStyle)
+    quote
+        $AtomicLevel{$(lvl.AVal), $(lvl.Lvl)}($(ctx(lvl.lvl)), $(lvl.locks))
+    end
+end
+
+function virtualize(ex, ::Type{AtomicLevel{AVal, Lvl}}, ctx, tag=:lvl) where {AVal, Lvl}
+    sym = freshen(ctx, tag)
+    atomics = freshen(ctx, tag, :_locks)
+    push!(ctx.preamble, quote
+            $sym = $ex
+            $atomics = $ex.locks
+        end)
+    lvl_2 = virtualize(:($sym.lvl), Lvl, ctx, sym)
+    temp = VirtualAtomicLevel(lvl_2, sym, atomics, typeof(level_default(Lvl)), Val, AVal, Lvl)
+    temp
+end
+
+Base.summary(lvl::VirtualAtomicLevel) = "Atomic($(lvl.Lvl))"
+virtual_level_resize!(lvl::VirtualAtomicLevel, ctx, dims...) = (lvl.lvl = virtual_level_resize!(lvl.lvl, ctx, dims...); lvl)
+virtual_level_size(lvl::VirtualAtomicLevel, ctx) = virtual_level_size(lvl.lvl, ctx)
+virtual_level_size(x, ctx) = error(string("Not defined for", x))
+virtual_level_eltype(lvl::VirtualAtomicLevel) = virtual_level_eltype(lvl.lvl)
+virtual_level_default(lvl::VirtualAtomicLevel) = virtual_level_default(lvl.lvl)
+
+function declare_level!(lvl::VirtualAtomicLevel, ctx, pos, init)
+    lvl.lvl = declare_level!(lvl.lvl, ctx, pos, init)
+    return lvl
+end
+
+function assemble_level!(lvl::VirtualAtomicLevel, ctx, pos_start, pos_stop)
+    pos_start = cache!(ctx, :pos_start, simplify(pos_start, ctx))
+    pos_stop = cache!(ctx, :pos_stop, simplify(pos_stop, ctx))
+    idx = freshen(ctx.code, :idx)
+    lockVal = freshen(ctx.code, :lock)
+    push!(ctx.code.preamble, quote 
+              Finch.resize_if_smaller!($(lvl.locks), $(ctx(pos_stop))) 
+              @inbounds for $idx = $(ctx(pos_start)):$(ctx(pos_stop))
+                lockVal = make_lock(eltype($(lvl.AVal)))
+                if !isnothing(lockVal)
```

Comment:
can we skip this isnothing and just set `make_lock` default to a pass-through? e.g. `make_lock(Bool) = false`

Reply:
`make_lock(Lock) = Lock()`